asim: use kvserver.LoadBasedRebalancingObjective

wenyihu6 opened the pull request:
Previously, LBRebalancingObjective was part of the simulation settings, using
QPS as the load rebalance objective. However, the cluster setting
kvserver.LoadBasedRebalancingObjective defaults to CPU rebalancing instead. This
commit updates asim to rely on the cluster setting, adopting the default cluster
setting configuration.

Epic: CRDB-55052
Release note: none 
Resolves: https://github.com/cockroachdb/cockroach/issues/154512

wenyihu6 committed:
asim: use kvserver.LoadBasedRebalancingObjective

Previously, LBRebalancingObjective was part of the simulation settings, using
QPS as the load rebalance objective. However, the cluster setting
kvserver.LoadBasedRebalancingObjective defaults to CPU rebalancing instead. This
commit updates asim to rely on the cluster setting, adopting the default cluster
setting configuration.

tbg reviewed (changes requested):
Thank you for doing this! Just a bit of test cleanup opportunity here before you merge.

> @tbg reviewed 8 of 8 files at r1, 1 of 1 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/155033#-:-Ob3r0Nk66LDWqZUEv7H:b949axv)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @wenyihu6)

___
*[`pkg/kv/kvserver/asim/tests/testdata/non_rand/mma/skewed_cpu_even_ranges_mma.txt` line 49 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/155033#-Ob3oaO8E8moQw0MTG1g:-Ob3oaO8E8moQw0MTG1h:b-mldcjr) ([raw file](https://github.com/cockroachdb/cockroach/blob/a84d45bdc5f1364d054a3d7f51be5455459626e2/pkg/kv/kvserver/asim/tests/testdata/non_rand/mma/skewed_cpu_even_ranges_mma.txt#L49)):*
> ```help
> cpu#1: last:  [s1=563639640, s2=579095747, s3=561019502, s4=561591114, s5=560289458, s6=563986913, s7=561058911, s8=559174716, s9=561248205] (stddev=5708749.39, mean=563456022.89, sum=5071104206)
> cpu#1: thrash_pct: [s1=61%, s2=93%, s3=85%, s4=33%, s5=25%, s6=31%, s7=26%, s8=31%, s9=38%]  (sum=423%)
> cpu_util#1: last:  [s1=0.11, s2=0.12, s3=0.11, s4=0.11, s5=0.11, s6=0.11, s7=0.11, s8=0.11, s9=0.11] (stddev=0.00, mean=0.11, sum=1)
> ```

nice and even :thumbsup: 
___
*[`pkg/kv/kvserver/asim/tests/testdata/non_rand/mma/skewed_cpu_even_ranges_mma.txt` line 56 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/155033#-Ob3olql-mwuKv8qHk3U:-Ob3olql-mwuKv8qHk3V:b-tz16sf) ([raw file](https://github.com/cockroachdb/cockroach/blob/a84d45bdc5f1364d054a3d7f51be5455459626e2/pkg/kv/kvserver/asim/tests/testdata/non_rand/mma/skewed_cpu_even_ranges_mma.txt#L56)):*
> ```help
> replicas#1: first: [s1=36, s2=36, s3=36, s4=36, s5=36, s6=36, s7=36, s8=36, s9=36] (stddev=0.00, mean=36.00, sum=324)
> replicas#1: last:  [s1=36, s2=38, s3=34, s4=36, s5=34, s6=36, s7=36, s8=38, s9=36] (stddev=1.33, mean=36.00, sum=324)
> replicas#1: thrash_pct: [s1=814%, s2=660%, s3=474%, s4=343%, s5=288%, s6=271%, s7=229%, s8=146%, s9=486%]  (sum=3711%)
> ```

nice and thrashy (good b/c this is sma)
___
*[`pkg/kv/kvserver/asim/tests/testdata/non_rand/sma/rebalancing.txt` line 3 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/155033#-Ob3qi2dAXP77dONekiA:-Ob3qi2dAXP77dONekiB:by3qqgw) ([raw file](https://github.com/cockroachdb/cockroach/blob/a84d45bdc5f1364d054a3d7f51be5455459626e2/pkg/kv/kvserver/asim/tests/testdata/non_rand/sma/rebalancing.txt#L3)):*
> ```help
> # This test verifies that the allocator can rebalance qps load across a 7-node
> # cluster. mma-only fails here since it doesn't balance based on leases.
> # TODO(tbg): this test works because the default rebalancing objective in asim
> ```

See this TODO. I'd say for now rename this file to `rebalancing_qps.txt` and change the cluster setting so that the old behavior is retained?
___
*[`pkg/kv/kvserver/asim/tests/testdata/non_rand/sma/multi_store_lease.txt` line 1 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/155033#-Ob3owVtBtGrptrKMih7:-Ob3owVtBtGrptrKMih8:b-cy8cm9) ([raw file](https://github.com/cockroachdb/cockroach/blob/a84d45bdc5f1364d054a3d7f51be5455459626e2/pkg/kv/kvserver/asim/tests/testdata/non_rand/sma/multi_store_lease.txt#L1)):*
> ```help
> # This test sets up a multi-store cluster.  We assert on this with a balance
> ```

I'm wondering what this test is adding. It looks like a pure write workload. SMA doesn't do anything with that, and so nothing happens. MMA will balance the write load (I'm guessing, the metric isn't printed here).

So this is really a pure write workload multi store test. So maybe we can take `high_write_uniform_cpu` and make it multi-store, then remove this test?


<!-- Sent from Reviewable.io -->

wenyihu6 committed:
asim: rename rebalancing.txt to rebalancing_qps.txt

This commit renames rebalancing.txt to rebalancing_qps.txt to clarify
that it performs rebalancing based on QPS.

wenyihu6 committed:
asim: make LoadBasedRebalancingObjective configurable

Previously, the simulator was updated to use cluster setting
LoadBasedRebalancingObjective, which defaults to CPU-based rebalancing. However,
the goal of rebalancing_qps.txt is to rebalance based on QPS. This commit
introduces LoadBasedRebalancingObjective as a configurable simulation setting,
allowing it to be adjusted via a cluster setting. And it updates rebalancing_qps
to explicitly set the rebalancing objective to QPS, restoring its original
intent after the recent simulator change.

wenyihu6 committed:
asim: clean up generated folder for rebalancing

This commit regenerates the rebalancing_qps generated files and removes
outdated generated data.

wenyihu6 committed:
asim: removes multi_store_lease.txt

This commit removes multi_store_lease.txt, as its purpose was unclear. The test
set up a multi-store cluster running a write workload, where sma rebalanced
based on leases and replicas, while mma additionally rebalanced on write
bandwidth. It only asserted that leases were evenly distributed. It's likely
that this test was originally added to increase multi-store coverage but did not
provide meaningful validation. To retain coverage, high_write_uniform_cpu.txt
has been updated to use a multi-store setup instead.

wenyihu6 reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/155033)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @tbg)

___
*[`pkg/kv/kvserver/asim/tests/testdata/non_rand/sma/multi_store_lease.txt` line 1 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/155033#-Ob3owVtBtGrptrKMih7:-ObD87DRDLeJKdAfaxUP:bri9ag0) ([raw file](https://github.com/cockroachdb/cockroach/blob/a84d45bdc5f1364d054a3d7f51be5455459626e2/pkg/kv/kvserver/asim/tests/testdata/non_rand/sma/multi_store_lease.txt#L1)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

I'm wondering what this test is adding. It looks like a pure write workload. SMA doesn't do anything with that, and so nothing happens. MMA will balance the write load (I'm guessing, the metric isn't printed here).

So this is really a pure write workload multi store test. So maybe we can take `high_write_uniform_cpu` and make it multi-store, then remove this test?
</blockquote></details>

Agreed it was likely added to get some multi-store coverage. Updated `high_write_uniform_cpu.txt` to use a multi-store setup and removed this test.

![Screenshot 2025-10-10 at 8.42.19 AM.png](https://files.reviewableusercontent.io/images/reviewable/56973754/YHnPMoRNmdQXaw0pSDlXX-TQ/Screenshot%202025-10-10%20at%208.42.19%E2%80%AFAM.png)
___
*[`pkg/kv/kvserver/asim/tests/testdata/non_rand/sma/rebalancing.txt` line 3 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/155033#-Ob3qi2dAXP77dONekiA:-ObD8_TQ8tnDqExtuIRX:b-s2rdzp) ([raw file](https://github.com/cockroachdb/cockroach/blob/a84d45bdc5f1364d054a3d7f51be5455459626e2/pkg/kv/kvserver/asim/tests/testdata/non_rand/sma/rebalancing.txt#L3)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

See this TODO. I'd say for now rename this file to `rebalancing_qps.txt` and change the cluster setting so that the old behavior is retained?
</blockquote></details>

Done.

```
❯ git --no-pager  diff head:pkg/kv/kvserver/asim/tests/testdata/non_rand/sma/rebalancing_qps.txt master:pkg/kv/kvserver/asim/tests/testdata/non_rand/sma/rebalancing.txt

diff --git a/pkg/kv/kvserver/asim/tests/testdata/non_rand/sma/rebalancing_qps.txt b/pkg/kv/kvserver/asim/tests/testdata/non_rand/sma/rebalancing.txt
index 98c490ba6c3..a872991433e 100644
--- a/pkg/kv/kvserver/asim/tests/testdata/non_rand/sma/rebalancing_qps.txt
+++ b/pkg/kv/kvserver/asim/tests/testdata/non_rand/sma/rebalancing.txt
@@ -1,5 +1,10 @@
 # This test verifies that the allocator can rebalance qps load across a 7-node
 # cluster. mma-only fails here since it doesn't balance based on leases.
+# TODO(tbg): this test works because the default rebalancing objective in asim
+# was accidentally left as 'qps' (even though we use 'cpu' in production).
+# When the default changes, this test needs to override the setting and probably
+# rename the test to specify that it's about qps rebalancing in particular,
+# which otherwise sees little explicit testing.
 skip_under_ci
 ----
 
@@ -30,11 +35,6 @@ assertion stat=qps type=steady ticks=6 upper_bound=0.05
 ----
 asserting: |qps(t)/mean_{T}(qps) - 1| ≤ 0.05 ∀ t∈T and each store (T=last 6 ticks)
 
-
-# Set rebalance objective to 0 to disable rebalancing.
-setting rebalance_objective=0
-----
-
 # The generators are then called and 2 simulation runs, named samples are
 # created and evaluated. Each sample has a fixed duration of 3 minutes.
 # Following the evaluation, the samples are checked individually against the
@@ -85,7 +85,6 @@ Cluster Set Up
 Key Space
        [0,10000): 7(rf=3), 0MiB, [s1:(7,4*),s2:(6,2*),s3:(3,0*),s4:(2,0*),s5:(1,0*),s6:(1,1*),s7:(1,0*)]
 Event
-       set LBRebalancingObjective to 0
        set LBRebalancingMode to 2
 Workload Set Up
        [0,10000): 95%r large-block [128-256B/op, 7000ops/s]
```


<!-- Sent from Reviewable.io -->

tbg reviewed (approved):
Thank you!

> @tbg reviewed 1 of 1 files at r3, 3 of 3 files at r4, 8 of 8 files at r5, 2 of 2 files at r6, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/155033#-:-OcjFvMJ0TCUe9BTBnXZ:b-9755fn)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @wenyihu6)



<!-- Sent from Reviewable.io -->

wenyihu6 commented:
TFTR!

bors r+
